### PR TITLE
Add 'triggerCharacter' option for completions requests

### LINKED ISSUE #21012: [TSServer] Completion trigger characters
**TL;DR**: Add the concept of a triggerCharacter for completions. Update the TS Server so that it uses the triggerCharacter to determine if completions should be returned at the requested location or not

## Background
VS Code uses trigger characters to automatically show suggestions as a user types. Registering `"` as a trigger character for example means that suggestions are shown automatically whenever a user types `"`. 

The VS Code TS extension uses these trigger character to determine if we should request completions from TS Server. For the trigger character `"` for example, the VS Code TS extension has logic so that it only requests completions when you start typing an import path:

```
import {} from "|
```

and not when you type out a string  #like:

```
const z = "abc"|
```

This is because the TS Server still returns completion entires in the second case, and we want to ensure these do not show up unless a user explicitly requests them.

The TS Server currently has no knowledge of what triggered a completions request.

## Problem
The VS Code TS extension's logic that determines when to request completions from the TSServer is very simple. It just runs some regular expressions on the current line to check if they resemble a valid auto-suggest context. This works well enough for limiting the `"` trigger to import paths, but there are other cases where this is too limiting.

Take:

```
const s: 'abc' | 'def' = "|
```

TS can provide helpful suggestions here but we do not show them automatically because triggering on every `"` would show suggestions in too many locations (like for the case `"abc"|` noted above).

Another example would be https://github.com/Microsoft/vscode/issues/40539 which tracks showing jsx tag suggestions automatically when you type `<`. Again, VS Code cannot just blindly register `<` as a trigger character, as this would cause suggestions to be shown whenever you type a less-than operator in an expression.

## Proposal
Make the TS server trigger character aware. Use the trigger character information and TypeScript's more complete understanding of the current document to determine when to return completions and when not to.

When the  trigger character is `"` for example, the TS server would return completions for import paths like:

```
import {} from "|
```

and for literal types like:

```
const s: 'abc' | 'def' = "|
```

but not for strings:

```
const z = "abc"|
```

A similar check could be done for the triggerCharacter `<`, which should only return completions when the `<` is used as a jsx tag opening.

To implement this, add a new optional `triggerCharacter` field to  `CompletionsRequestArgs `

```ts
interface CompletionsRequestArgs extends FileLocationRequestArgs {
    [...]

    triggerCharacter?: string;
}
```

When a triggerCharacter is provided, use this to determine if completions should be returned or not. If the triggerCharacter is `undefined`, there would be no change from the existing behavior

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #21012

@mjbvz I don't think it should be necessary for the editor to send us the trigger character? Shouldn't that always be the character immediately preceding `position`?
@amcasey You may be interested in using this in visual studio too.

### CODE REVIEW

**Comment by mjbvz** [comment]
@andy-ms Yes I would expect that it should be but I don't know for sure that this is correct 100% of the time

**Comment by andy-ms** [comment]
If it isn't correct, the code I added here will have an assertion failure, so I guess we'd find out..

**Comment by mhegazy** [comment]
> I don't think it should be necessary for the editor to send us the trigger character? Shouldn't that always be the character immediately preceding position?

We need to distinguish between the cases where the user explicitly asks for completion (ctrl + space) vs when the editor triggers it automatically. the user could trigger completion using ctrl+space after a closing `"` for example. We could just say we will not support that. i.e. both cases get no completions..

File: src/services/completions.ts [hunk 1]
```diff
@@ -2197,4 +2198,27 @@ namespace ts.Completions {
...
+            case "'":
+            case "`":
+                switch (contextToken.kind) {
+                    case SyntaxKind.StringLiteral:
+                    case SyntaxKind.NoSubstitutionTemplateLiteral:
+                    case SyntaxKind.TemplateExpression:
+                    case SyntaxKind.TaggedTemplateExpression:
+                        break;
+                    default:
+                        Debug.fail();
```
Comment (mhegazy) [inline src/services/completions.ts]: might wanna add the contextToken to the error message

Comment (mhegazy) [reply src/services/completions.ts]: the quote could be in a comment, so not sure you want to fail always here.

File: src/services/completions.ts [hunk 2]
```diff
@@ -2197,4 +2198,27 @@ namespace ts.Completions {
...
+                }
+                return position === contextToken.parent.end;
+            case "<":
+                Debug.assert(contextToken.kind === SyntaxKind.LessThanToken);
+                return contextToken.parent.kind === SyntaxKind.BinaryExpression;
```
Comment (DanielRosenwasser) [inline src/services/completions.ts]: There's technically no reason why `<` couldn't trigger type completions when the LHS is a type or has generic signatures.

File: src/services/completions.ts [hunk 3]
```diff
@@ -2197,4 +2198,28 @@ namespace ts.Completions {
...
+            case '"':
+            case "'":
+            case "`":
+                switch (contextToken.kind) {
+                    case SyntaxKind.StringLiteral:
+                    case SyntaxKind.NoSubstitutionTemplateLiteral:
+                    case SyntaxKind.TemplateExpression:
+                    case SyntaxKind.TaggedTemplateExpression:
+                        // Don't automatically bring up completions at closing quote
+                        return position === contextToken.parent.end;
```
Comment (mhegazy) [inline src/services/completions.ts]: not to be pedantic, but you might want to check the first char in the string to match the trigger char to allow for things like:
```ts
var x: "`a`" | "`b`" = "`
```

File: src/services/completions.ts [hunk 4]
```diff
@@ -2197,4 +2198,28 @@ namespace ts.Completions {
...
+                    default:
+                        if (!isInComment(sourceFile, position, contextToken)) Debug.failBadSyntaxKind(contextToken);
```
Comment (mhegazy) [inline src/services/completions.ts]: not sure you need this. We are doing some extra work just to know it is invalid. i would just return directly and avoid walking the tree every time in a comment just for an assert.

File: src/services/completions.ts [hunk 5]
```diff
@@ -2197,4 +2198,28 @@ namespace ts.Completions {
     function hasIndexSignature(type: Type): boolean {
         return !!type.getStringIndexType() || !!type.getNumberIndexType();
     }
+
+    function isInvalidTrigger(sourceFile: SourceFile, triggerCharacter: string | undefined, contextToken: Node, position: number): boolean {
```
Comment (mhegazy) [inline src/services/completions.ts]: consider switching this to `isValidTrigger`, and changing the condition to be `if (triggerCharacter && !isValid()) return undefined`

File: src/services/completions.ts [hunk 6]
```diff
@@ -2197,4 +2198,31 @@ namespace ts.Completions {
...
+
+    function isValidTrigger(sourceFile: SourceFile, triggerCharacter: string, contextToken: Node, position: number): boolean {
+        switch (triggerCharacter) {
+            case '"':
+            case "'":
+            case "`":
+                // Only automatically bring up completions if this is an opening quote.
+                return isStringLiteralOrTemplate(contextToken) && position === contextToken.getStart(sourceFile) + 1;
+            case "<":
+                Debug.assert(contextToken.kind === SyntaxKind.LessThanToken);
```
Comment (mhegazy) [inline src/services/completions.ts]: if this is inside a comment it is not going be a lessThanToken, right?

Comment (mhegazy) [reply src/services/completions.ts]: Also add tests for both kinds of triggers in comments and in strings

**Comment by amcasey** [comment]
Presently, VS does not use quotes as trigger characters so it has no need to disambiguate start from end quotes when triggering completion.  However, it's nice to know that this functionality could be lit up if someone asked for it.

**Pull request merged**